site/data/nextdoor.ts

```diff
@@ -1,8 +1,115 @@
+import { getParameter, putParameter } from "@create-disruptions-data/shared-ts/utils/ssm";
+import { addSocialAccountToOrg, getOrgSocialAccounts, upsertSocialMediaPost } from "./dynamo";
+import { NEXTDOOR_AUTH_URL, NEXTDOOR_URL } from "../constants";
+import {
+    GroupIds,
+    nextdoorAgencyBoundaryResultSchema,
+    nextdoorMeSchema,
+    nextdoorTokenSchema,
+} from "../schemas/nextdoor.schema";
 import { SocialMediaAccount } from "../schemas/social-media-accounts.schema";
+import { NextdoorPost } from "../schemas/social-media.schema";
+import logger from "../utils/logger";
 
-export const getNextdoorAuthUrl = () => {
-    return "";
+export const nextdoorRedirectUri = `${process.env.DOMAIN_NAME as string}/api/nextdoor-callback`;
+
+export const getNextdoorAuthHeader = async () => {
+    const { nextdoorClientId, nextdoorClientSecret } = await getNextdoorClientIdAndSecret();
+    const key = `${nextdoorClientId}:${nextdoorClientSecret}`;
+
+    return `Basic ${Buffer.from(key).toString("base64")}`;
+};
+
+export const getNextdoorSsmKey = (orgId: string) => `/social/nextdoor/${orgId}/refresh_token`;
+
+export const getNextdoorAccessToken = async (orgId: string) => {
+    const refreshTokenParam = await getParameter(getNextdoorSsmKey(orgId), logger, true);
+
+    if (!refreshTokenParam.Parameter?.Value) {
+        throw new Error("Refresh token not found");
+    }
+
+    const refreshToken = refreshTokenParam.Parameter.Value;
+
+    return refreshToken;
+};
+
+export const addNextdoorAccount = async (
+    code: string,
+    orgId: string,
+    addedBy: string,
+    createdByOperatorOrgId?: string | null,
+) => {
+    const authHeader = await getNextdoorAuthHeader();
+
+    const tokenResponse = await fetch(`${NEXTDOOR_AUTH_URL}v2/token`, {
+        method: "POST",
+        body: new URLSearchParams({
+            grant_type: "authorization_code",
+            code,
+            redirect_uri: nextdoorRedirectUri,
+        }),
+        headers: {
+            "Content-Type": "application/x-www-form-urlencoded",
+            Authorization: authHeader,
+        },
+    });
+
+    if (!tokenResponse.ok) {
+        const message = `An error has occurred whilst authenticating Nextdoor user: ${tokenResponse.status}`;
+        throw new Error(message);
+    }
+
+    const tokenResult = nextdoorTokenSchema.parse(await tokenResponse.json());
+
+    const userDetailsResponse = await fetch(`${NEXTDOOR_URL}external/api/partner/v1/me/`, {
+        method: "GET",
+        headers: {
+            Authorization: `Bearer ${tokenResult.accessToken}`,
+            Accept: "application/json",
+        },
+    });
+
+    if (!userDetailsResponse.ok) {
+        const message = `An error has occurred retrieving account information: ${userDetailsResponse.status}`;
+        throw new Error(message);
+    }
+
+    const userDetails = nextdoorMeSchema.parse(await userDetailsResponse.json());
```

Comment:
Could we raise this with Emily?

Reply:
Had a chat with Emily and we agreed on throwing an error on the social media accounts page if someone tries to sign up with an account that isn't an agency account type.

Also, as part of that chat we thought it best to only allow org admins to add nextdoor accounts. So operator users should not be able to connect a nextdoor account. This is because we don't know what nextdoor account type an operator has and we're essentially just building this feature for one LTA user.